### ISSUE TITLE
Reimplement tosa.rescale/tosa.apply_scale lowering to not use 64-bit math and vectorize.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#7781]
We ran into an issue with an i8 quantized model (https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/mobilenet_v2_1.0_224_quant.tgz) that was trying to use 64-bit math in VMVX (not supported by default). This highlighted that we are still doing the full expansion of rescale ops to non-trivial 64-bit math:

```mlir
%145 = "tosa.rescale"(%144) {double_round = true, input_zp = 0 : i32, multiplier = [1764866200 : i32], output_zp = -128 : i32, per_channel = false, scale32 = true, shift = [36 : i32]} : (tensor<1x7x7x1280xi32>) -> tensor<1x7x7x1280xi8>
```

```mlir
      func @main_dispatch_72(%arg0: !stream.binding {stream.alignment = 32 : index}, %arg1: !stream.binding {stream.alignment = 32 : index}) {
            ...
            scf.for %arg4 = %6 to %c1280 step %7 {
              %8 = affine.min #map56(%arg2)[%workgroup_size_2]
              %9 = affine.min #map56(%arg3)[%workgroup_size_1]
              %10 = affine.min #map53(%arg4)[%workgroup_size_0]
              %11 = flow.dispatch.tensor.load %0, offsets = [0, %arg2, %arg3, %arg4], sizes = [1, %8, %9, %10], strides = [1, 1, 1, 1] : !flow.dispatch.tensor<readonly:1x1x1x1280xi32> -> tensor<1x?x?x?xi32>
              %12 = linalg.init_tensor [1, %8, %9, %10] : tensor<1x?x?x?xi8>
              %13 = linalg.generic {indexing_maps = [#map12, #map12], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%11 : tensor<1x?x?x?xi32>) outs(%12 : tensor<1x?x?x?xi8>) {
              ^bb0(%arg5: i32, %arg6: i8):  // no predecessors
                %14 = arith.subi %arg5, %c-6272_i32 : i32
                %15 = "tosa.apply_scale"(%14, %c21913098_i32, %c30_i8) {double_round = false} : (i32, i32, i8) -> i32
                %16 = arith.addi %15, %c-128_i32 : i32
                %17 = arith.cmpi slt, %16, %c-128_i32 : i32
                %18 = select %17, %c-128_i32, %16 : i32
                %19 = arith.cmpi slt, %c127_i32, %16 : i32
                %20 = select %19, %c127_i32, %18 : i32
                %21 = arith.trunci %20 : i32 to i8
                linalg.yield %21 : i8
              } -> tensor<1x?x?x?xi8>
              flow.dispatch.tensor.store %13, %1, offsets = [0, %arg2, %arg3, %arg4], sizes = [1, %8, %9, %10], strides = [1, 1, 1, 1] : tensor<1x?x?x?xi8> -> !flow.dispatch.tensor<writeonly:1x1x1x1280xi8>
            }
          }
```
(note that the apply_scale survives long into translation)

->

```mlir
        llvm.func internal @main_dispatch_72(%arg0: !llvm.ptr<struct<"iree_hal_executable_dispatch_state_v0_t", (array<3 x i32>, array<3 x i32>, i64, ptr<i32>, i64, ptr<ptr<i8>>, ptr<i64>, ptr<func<i32 (ptr<func<i32 (ptr<i8>)>>, ptr<i8>)>>, ptr<ptr<func<i32 (ptr<i8>)>>>)>>, %arg1: !llvm.ptr<array<3 x i32>>, %arg2: !llvm.ptr<i8>) -> i32 attributes {sym_visibility = "private"} {
         ...
        ^bb7:  // pred: ^bb6
          %63 = llvm.mul %1, %2  : i64
          %64 = llvm.add %59, %63  : i64
          %65 = llvm.add %64, %63  : i64
          %66 = llvm.add %65, %63  : i64
          %67 = llvm.add %66, %61  : i64
          %68 = llvm.getelementptr %51[%67] : (!llvm.ptr<i32>, i64) -> !llvm.ptr<i32>
          %69 = llvm.load %68 : !llvm.ptr<i32>
          %70 = llvm.sub %69, %4  : i32
          %71 = llvm.mlir.constant(29 : i8) : i8
          %72 = llvm.sext %71 : i8 to i64
          %73 = llvm.shl %17, %72  : i64
          %74 = llvm.sext %70 : i32 to i64
          %75 = llvm.mlir.constant(21913098 : i64) : i64
          %76 = llvm.mlir.constant(30 : i64) : i64
          %77 = llvm.mul %74, %75  : i64
          %78 = llvm.add %77, %73  : i64
          %79 = llvm.ashr %78, %76  : i64
          %80 = llvm.trunc %79 : i64 to i32
          %81 = llvm.add %80, %5  : i32
          %82 = llvm.icmp "slt" %81, %5 : i32
          %83 = llvm.select %82, %5, %81 : i1, i32
          %84 = llvm.icmp "slt" %6, %81 : i32
          %85 = llvm.select %84, %6, %83 : i1, i32
          %86 = llvm.trunc %85 : i32 to i8
          %87 = llvm.getelementptr %60[%67] : (!llvm.ptr<i8>, i64) -> !llvm.ptr<i8>
          llvm.store %86, %87 : !llvm.ptr<i8>
          %88 = llvm.add %61, %3  : i64
          llvm.br ^bb6(%88 : i64)
          ...
```

Going i8 -> i64 -> i8 is excessive but more importantly on 32-bit platforms (including mobile GPUs) this will be emulated math and much slower than it needs to be. Even on 64-bit platforms this (if it vectorized, which it doesn't) would effectively halve the amount of work we can do by using up twice as much register space. According to https://discord.com/channels/689900678990135345/804059051750457404/915274606837903380 tflite and the ARM compute library both use i32 math and we should do the same.

Here's the aarch64 intrinsic version using 32-bit math: https://github.com/ARM-software/ComputeLibrary/blob/7dcb9fadb98cad05fca72de3273311d570d98b4e/src/core/NEON/kernels/arm_gemm/quantized.cpp
This shows roughly the mix of instructions we should be ending up with and a benchmark for how close we are to doing that. Today we are quite far:

```asm
.LBB65_16:
	.loc	2 254 12
	ldr	q5, [x23, x26]
	dup	v6.2d, x6
	add	x26, x26, #16
	mov	v7.16b, v6.16b
	add	v5.4s, v5.4s, v4.4s
	ext	v16.16b, v5.16b, v5.16b, #8
	smlal	v6.2d, v5.2s, v1.2s
	smlal	v7.2d, v16.2s, v1.2s
	shrn	v5.2s, v6.2d, #30
	shrn2	v5.4s, v7.2d, #30
	smax	v5.4s, v5.4s, v0.4s
	umin	v5.4s, v5.4s, v2.4s
	xtn	v5.4h, v5.4s
	eor	v5.8b, v5.8b, v3.8b
	xtn	v5.8b, v5.8h
	str	s5, [x24, x27]
	add	x27, x27, #4
	cmp	x27, #64
```

**Comment by benvanik** [comment]
oh, and this matters because in a quantized model nearly every single kernel ends up with a rescaling operation, meaning that we are likely running nearly everything in this degraded state:
![image](https://user-images.githubusercontent.com/75337/144089843-05f9bf75-ddc2-4117-9b99-8b575369fcec.png)

**Comment by benvanik** [comment]
Here's some additional information from MatthewARM:
> I've found out a bit more - George at Arm tells me that the int64 is unavoidable in reference code, but the optimised Arm code avoids it with the sqrdmul instruction

tflite reference:
https://github.com/tensorflow/tensorflow/pull/50290/files

I vaguely recall us talking about a better lowering path than having apply_scale tunnel all the way down so that we could use the instructions and avoid the extra work but that seems to have not been done and only tosa->std exists:
https://github.com/llvm/llvm-project/blob/a54f4eae0e1d0ef5adccdcf9f6c2b518dc1101aa/mlir/lib/Conversion/TosaToStandard/TosaToStandard.cpp#L58

Would be good to evaluate how to better lower that to vector dialect so that we can get the proper fusion/folding with that as we switch off of LLVM autovectorization.

**Comment by benvanik** [comment]
If we need an op in the VM to do this I can add it - or emulate it - but not having to add a tosa.apply_rescale and instead using something standard would be ideal.

**Comment by MaheshRavishankar** [comment]
The best way forward I see is to not promote to 64-bit to begin with. This was done to match the reference implementation I think. For the quantized path, thats overly conservative? Would be really good to just resolve this at TOSA level without it even surviving to IREE core?

**Comment by benvanik** [comment]
agreed that'd be ideal - it should just be math - I think the issue was that the saturating multiply had no representation in arith so no way to get that lowered through without emulating it (as it does with those compares) - but this may just mean a single arith op like fma for multiply w/ saturation and then we're done

**Comment by benvanik** [comment]
There's also some silly cases that show up with double-rescales that don't seem to get cleaned up - moving to more standard ops will hopefully let some of that fall out but we may want tosa-level optimizations as well:
```mlir
%166 = "tosa.rescale"(%165) {double_round = true, input_zp = 0 : i32, multiplier = [1257485775 : i32], output_zp = -19 : i32, per_channel = false, scale32 = true, shift = [39 : i32]} : (tensor<1x14x14x64xi32>) -> tensor<1x14x14x64xi8>
    %167 = "tosa.rescale"(%166) {double_round = false, input_zp = -19 : i32, multiplier = [1999568813 : i32], output_zp = 0 : i32, per_channel = false, scale32 = true, shift = [12 : i32]} : (tensor<1x14x14x64xi8>) -> tensor<1x14x14x64xi32>
```

**Comment by benvanik** [comment]
It also looks like there are many cases where it goes rescale->[ conv | rescale ], which causes the rescales to be in their own dispatch region today - ensuring they are always fused with producers (when narrowing) and consumers (when widening) is going to be critical (right now we do things like rescale i8->i32 into memory and then read those i32 in the consumers) - maybe something that is easier to tackle once this is all represented right.

**Comment by MaheshRavishankar** [comment]
I cant see the types in the snippet you have, but rescale -> conv needs to be represented as a new named op.

**Comment by MaheshRavishankar** [comment]
> agreed that'd be ideal - it should just be math - I think the issue was that the saturating multiply had no representation in arith so no way to get that lowered through without emulating it (as it does with those compares) - but this may just mean a single arith op like fma for multiply w/ saturation and then we're done

yeah we probably need to add more ops to `arith` dialect for that. That is the easy part.

**Comment by benvanik** [comment]
Ahh we may need a few of those named ops then - I think the important thing is that we aren't materializing into memory in unquantized/widened values, but also want to ensure that within a dispatch we aren't rescaling multiple times - the composition of that with linalg -> dispatch region formation -> loops escapes me though :P

**Comment by benvanik** [comment]
do you think we'll still be able to make the guarantees that across dispatch regions we aren't materializing full unquantized values into memory if we remove all of the semantics? (I could see ways to try to ensure the I/O edges are on narrow/widen, but in my mind the solutions to implement that are all hacky)

**Comment by benvanik** [comment]
Patterns like this are what we see a lot of:
```mlir
  %210 = "tosa.depthwise_conv2d"(%209, %101, %62) {dilation = [1, 1], pad = [1, 1, 1, 1], quantization_info = {input_zp = -128 : i32, weight_zp = 31 : i32}, stride = [1, 1]} : (tensor<1x14x14x576xi8>, tensor<3x3x576x1xi8>, tensor<576xi32>) -> tensor<1x14x14x576xi32>
  %211 = "tosa.rescale"(%210) {double_round = true, input_zp = 0 : i32, multiplier = [1279693056 : i32], output_zp = -128 : i32, per_channel = false, scale32 = true, shift = [34 : i32]} : (tensor<1x14x14x576xi32>) -> tensor<1x14x14x576xi8>
  %212 = "tosa.conv2d"(%211, %63, %64) {dilation = [1, 1], pad = [0, 0, 0, 0], quantization_info = {input_zp = -128 : i32, weight_zp = 26 : i32}, stride = [1, 1]} : (tensor<1x14x14x576xi8>, tensor<96x1x1x576xi8>, tensor<96xi32>) -> tensor<1x14x14x96xi32>
  %213 = "tosa.rescale"(%212) {double_round = true, input_zp = 0 : i32, multiplier = [1666054802 : i32], output_zp = -1 : i32, per_channel = false, scale32 = true, shift = [39 : i32]} : (tensor<1x14x14x96xi32>) -> tensor<1x14x14x96xi8>
  %214 = "tosa.rescale"(%213) {double_round = false, input_zp = -1 : i32, multiplier = [1073741824 : i32], output_zp = 0 : i32, per_channel = false, scale32 = true, shift = [11 : i32]} : (tensor<1x14x14x96xi8>) -> tensor<1x14x14x96xi32>
  %215 = "tosa.rescale"(%207) {double_round = false, input_zp = -1 : i32, multiplier = [2031709950 : i32], output_zp = 0 : i32, per_channel = false, scale32 = true, shift = [12 : i32]} : (tensor<1x14x14x96xi8>) -> tensor<1x14x14x96xi32>
  %216 = "tosa.add"(%214, %215) : (tensor<1x14x14x96xi32>, tensor<1x14x14x96xi32>) -> tensor<1x14x14x96xi32>
  %217 = "tosa.rescale"(%216) {double_round = true, input_zp = 0 : i32, multiplier = [1713159654 : i32], output_zp = -2 : i32, per_channel = false, scale32 = true, shift = [50 : i32]} : (tensor<1x14x14x96xi32>) -> tensor<1x14x14x96xi8>
```

if the depthwise_conv2d and conv2d end up in different regions we want them to have the rescale edges, and if the add gets fused into the conv2d (which it should) we want the intermediate rescales to be elided

**Comment by MaheshRavishankar** [comment]
I am not saying we remove the semantics. The semantics should be captured using named ops. Potentially also need some `arith` operations that do the special math. That is the only way I see possible.

**Comment by benvanik** [comment]
ah ok didn't think of named arith ops - perhaps attributes are another route to avoid needing to have 1:1 copies of things - but if you see a path that's awesome :)

**Comment by MaheshRavishankar** [comment]
> if the depthwise_conv2d and conv2d end up in different regions we want them to have the rescale edges, and if the add gets fused into the conv2d (which it should) we want the intermediate rescales to be elided

If the `rescale` is combined with the `conv2d`, the edge carries the information in types?

**Comment by benvanik** [comment]
it currently falls over today if the rescales have multiple uses and such - but hopefully not a limitation forever!

**Comment by MaheshRavishankar** [comment]
thats a separate issue. I am saying `tosa.rescale -> `tosa.conv2d` is instead `tosa.conv2d_some_type_input_some_type_output`.

**Comment by rsuderman** [comment]
Many of the tosa.rescales can be optimized away (on my list) so I wouldn't focus on them. Many are actually no-ops or broadcasts.

As for `tosa.rescale`, the important focus is on `tosa.apply_scale`: due to how tosa performs its rescale it is guaranteed we either need some custom arith-ops that do mul-add-shift (iirc) OR we need to do the math in i64. It would be valuable to have some `arith` operators to prevent this but I am not the expert on what those ops should look like.

**Comment by benvanik** [comment]
We may not want to get rid of all the rescales to early, though - that's the compositional complexity here and what Mahesh's proposal may help with. If we got rid of the information encoded in the IR that a value can be represented in i8 we can't get that back later. I believe it's only after dispatch region formation we'd want to fold away the rescaling math - assuming we don't do silly things (like we do today) like putting rescales in their own dispatch regions and such - also something Mahesh's proposal should help with (a named conv_with_rescale can't be split apart too early).

**Comment by dcaballe** [comment]
We do vectorize `tosa.apply_scale` in Linalg and then lower it to a vector native or emulated 64-bit integer implementation when lowering to LLVM

**Comment by MaheshRavishankar** [comment]
Dropping codegen label for now based on above reply. Will let Rob close it out.

**Comment by rsuderman** [comment]
Completed. There is now an optional pass that lowers to purely 32-bit operations. Used for the RISC-V lowerings.
